Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp

```diff
@@ -313,68 +324,134 @@ bool MediaPlayerPrivateGStreamerBase::handleSyncMessage(GstMessage* message)
+                    m_initDatas.add(eventKeySystemIdString, WTFMove(initDataVector));
 #endif
-            // Here we receive the DRM init data from the pipeline: we will emit
-            // the needkey event with that data and the browser might create a
-            // CDMSession from this event handler. If such a session was created
-            // We will emit the message event from the session to provide the
-            // DRM challenge to the browser and wait for an update. If on the
-            // contrary no session was created we won't wait and let the pipeline
-            // error out by itself.
-            GstBuffer* data;
-            const char* keySystemId;
-            gboolean valid = gst_structure_get(structure, "data", GST_TYPE_BUFFER, &data,
-                                               "key-system-id", G_TYPE_STRING, &keySystemId, nullptr);
-            GstMapInfo mapInfo;
-            if (!valid || !gst_buffer_map(data, &mapInfo, GST_MAP_READ))
+                    m_handledProtectionEvents.add(GST_EVENT_SEQNUM(event.get()));
+                }
+                gst_buffer_unmap(data, &mapInfo);
+            }
+
+            if (catenatedInitDataChunksNumber == 0)
                 return false;
 
-            GST_DEBUG("scheduling keyNeeded event");
-            // FIXME: Provide a somehow valid sessionId.
+            if (catenatedInitDataChunksNumber > 1)
+                eventKeySystemIdString = emptyString();
+
+            RunLoop::main().dispatch([this, eventKeySystemIdString, initData = WTFMove(catenatedInitDataChunks)] {
+                GST_DEBUG("scheduling keyNeeded event for %s with catenated init datas size of %u", eventKeySystemIdString.utf8().data(), initData->size());
+                GST_MEMDUMP("init datas", reinterpret_cast<const unsigned char *>(initData->data()), initData->size());
+
+                // FIXME: Provide a somehow valid sessionId.
 #if ENABLE(ENCRYPTED_MEDIA)
-            needKey(keySystemId, "sessionId", reinterpret_cast<const unsigned char *>(mapInfo.data), mapInfo.size);
+                needKey(keySystemUuidToId(eventKeySystemIdString).string(), "sessionId", reinterpret_cast<const unsigned char *>(initData->data()), initData->size());
 #elif ENABLE(ENCRYPTED_MEDIA_V2)
-            RefPtr<Uint8Array> initData = Uint8Array::create(reinterpret_cast<const unsigned char *>(mapInfo.data), mapInfo.size);
-            needKey(initData);
+                RefPtr<Uint8Array> initDataArray = Uint8Array::create(reinterpret_cast<const unsigned char *>(initData->data()), initData->size());
+                needKey(initDataArray);
 #else
-            ASSERT_NOT_REACHED();
+                ASSERT_NOT_REACHED();
 #endif
-            gst_buffer_unmap(data, &mapInfo);
+            });
+
+            GST_INFO("waiting for a key request to arrive");
+            LockHolder lock(m_protectionMutex);
+            m_protectionCondition.waitForSeconds(m_protectionMutex, 4);
```

Comment:
Yes, it works, when I read the code at the beginning I had the impression that it would wait forever but I was clearly wrong and actually this works much better.